google-beta/resource_compute_snapshot.go

```diff
@@ -1,220 +1,536 @@
+// ----------------------------------------------------------------------------
+//
+//     ***     AUTO GENERATED CODE    ***    AUTO GENERATED CODE     ***
+//
+// ----------------------------------------------------------------------------
+//
+//     This file is automatically generated by Magic Modules and manual
+//     changes will be clobbered when the file is regenerated.
+//
+//     Please read more about how to change this file in
+//     .github/CONTRIBUTING.md.
+//
+// ----------------------------------------------------------------------------
+
 package google
 
 import (
 	"fmt"
 	"log"
+	"reflect"
+	"strconv"
 	"time"
 
+	"github.com/hashicorp/terraform/helper/customdiff"
 	"github.com/hashicorp/terraform/helper/schema"
-	"google.golang.org/api/compute/v1"
-	"google.golang.org/api/googleapi"
+	compute "google.golang.org/api/compute/v1"
 )
 
+func customDiffComputeSnapshotSnapshotEncryptionKeys(diff *schema.ResourceDiff, meta interface{}) error {
+	// oC -> Old Convenience Value
+	// nC -> New Convenience Value
+	// oN -> Old New Field Value
+	// nN -> New New Field Value
+	oC, nC := diff.GetChange("snapshot_encryption_key_raw")
+	oN, nN := diff.GetChange("snapshot_encryption_key.0.raw_key")
+
+	if nC != "" && nN != "" {
+		return fmt.Errorf("can't use snapshot_encryption_key_raw and snapshot_encryption_key.0.raw_key at the same time." +
+			"If you're removing snapshot_encryption_key.0.raw_key, set the value to \"\" instead. This is due to limitations in terraform plan.")
+	}
+
+	// Either field (convenience or new) has a value
+	// and then has another different value, so we ForceNew.
+	// We need to handle _EVERY_ ForceNew case in this diff
```

Comment:
Just commenting here to say this is the correct way to handle conditional ForceNew.

Reply:
Could all of this be turned into a helper function (`customDiffComputeSnapshotSnapshotEncryptionKeysForceNew`?) that accepts the old and new values, and returns the field that is ForceNew as a string, or `""` if no ForceNew is needed? I only ask because then we could do a table-driven unit test to make sure we're hitting all the edge cases without needing the expensive Terraform overhead. Just a thought, not a requirement.

Reply:
I think it could, but I'm not sure it's worth the effort. Covering the cases correctly isn't a big concern of mine, and the Terraform-level tests already do that; it's handling weirdness from how Terraform does diffs correctly.

Reply:
👍 Fair, wasn't sure where the concern around this function was. If you think the regular tests have this under control and are more worried about the diffing behaviour itself, agreed this is unnecessary.